pubkey/src/lib.rs

```diff
@@ -158,10 +160,133 @@ impl From<u64> for PubkeyError {
+    #[derive(Clone)]
+    pub struct PubkeyHasherBuilder {
+        offset: usize,
+    }
+
+    impl Default for PubkeyHasherBuilder {
+        fn default() -> Self {
+            std::thread_local!(static OFFSET: core::cell::Cell<usize>  = {
+                let mut rng = rand::prelude::thread_rng();
+                core::cell::Cell::new(rng.gen_range(0..32 - 8))
+            });
+
+            let offset = OFFSET.with(|offset| {
+                let mut next_offset = offset.get() + 1;
+                if next_offset > 24 {
+                    next_offset = 0;
+                }
+                offset.set(next_offset);
+                next_offset
+            });
+            PubkeyHasherBuilder { offset }
+        }
+    }
+
+    impl std::hash::BuildHasher for PubkeyHasherBuilder {
+        type Hasher = PubkeyHasher;
+        #[inline]
+        fn build_hasher(&self) -> Self::Hasher {
+            PubkeyHasher {
+                offset: self.offset,
+                state: 0,
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use {
+            super::PubkeyHasherBuilder,
+            crate::Pubkey,
+            core::hash::{BuildHasher, Hasher},
+        };
+        #[test]
+        fn test_pubkey_hasher_builder() {
+            let key = Pubkey::new_unique();
+            let builder = PubkeyHasherBuilder::default();
+            let mut hasher1 = builder.build_hasher();
+            let mut hasher2 = builder.build_hasher();
+            hasher1.write(key.as_array());
+            hasher2.write(key.as_array());
+            assert_eq!(
+                hasher1.finish(),
+                hasher2.finish(),
+                "Hashers made with same builder should be identical"
+            );
+            // Make sure that when we make new builders we get different slices
+            // chosen for hashing
+            let builder2 = PubkeyHasherBuilder::default();
+            for _ in 0..64 {
+                let mut hasher3 = builder2.build_hasher();
+                hasher3.write(key.as_array());
+                std::dbg!(hasher1.finish());
+                std::dbg!(hasher3.finish());
+                if hasher1.finish() != hasher3.finish() {
+                    return;
+                }
+            }
+            panic!("Hashers built with different builder should be different due to random offset");
+        }
+
+        #[test]
+        fn test_pubkey_hasher() {
+            let key1 = Pubkey::new_unique();
+            let key2 = Pubkey::new_unique();
+            let builder = PubkeyHasherBuilder::default();
+            let mut hasher1 = builder.build_hasher();
+            let mut hasher2 = builder.build_hasher();
+            hasher1.write(key1.as_array());
+            hasher2.write(key2.as_array());
+            assert_ne!(hasher1.finish(), hasher2.finish());
+        }
+    }
+}
+#[cfg(all(feature = "rand", not(target_os = "solana")))]
+pub use hasher::PubkeyHasherBuilder;
```

Comment:
Sorry if this is a silly question, but shouldn't this also need to make `PubkeyHasher` available?

Reply:
Actually we do not have to, but there is no harm in making it visible.

Reply:
Weird, how does that work then? I thought creating the hasher would give a "pub function returns a private type" error